src/classes/BGE_BatchGiftEntry_UTIL.cls

```diff
@@ -73,87 +73,84 @@ public with sharing class BGE_BatchGiftEntry_UTIL {
+                BDI_MappingServiceAdvanced.getInstance();
+
+        Map<String, BDI_FieldMapping> fieldMappingByDevName =
+                mappingService.fieldMappingByDevName;
 
-            Map<String, BDI_FieldMapping> fieldMappingByDevName =
-                    mappingService.fieldMappingByDevName;
+        for (GE_Template.Section formSection : deserializeTemplate(batch).layout.sections) {
+            for (GE_Template.Element formElement : formSection.elements) {
+                if (formElement.dataImportFieldMappingDevNames == null) {
+                    continue;
+                }
+                // Get the field mappings, used to reference the source fields
+                for (String fieldMappingName :
+                        formElement.dataImportFieldMappingDevNames) {
+
+                    BDI_FieldMapping fieldMapping =
+                            fieldMappingByDevName.get(fieldMappingName);
 
-            GE_Template.Template template = (GE_Template.Template) JSON.deserialize(
-                    batch.Form_Template__r.Template_JSON__c,
-                    GE_Template.Template.class
-            );
+                    Boolean isObjectMappedLookup = mappingService
+                            .importedRecordFieldNames.contains(
+                            fieldMappingName.toLowerCase());
 
-            for (GE_Template.Section formSection : template.layout.sections) {
-                for (GE_Template.Element formElement : formSection.elements) {
-                    if (formElement.dataImportFieldMappingDevNames == null) {
+                    if (fieldMapping == null && !isObjectMappedLookup) {
                         continue;
                     }
-                    // Get the field mappings, used to reference the source fields
-                    for (String fieldMappingName :
-                            formElement.dataImportFieldMappingDevNames) {
+                    String sourceFieldName = fieldMapping != null ?
+                            fieldMapping.Source_Field_API_Name :
+                            isObjectMappedLookup ?
+                                    fieldMappingName : null;
+                    SObjectField field = DataImport__c.SObjectType.getDescribe()
+                            .fields.getMap().get(sourceFieldName);
+
+                    // Add the source fields used in the Template but only if the source field name is valid
+                    if (!fieldApiNames.contains(sourceFieldName.toLowerCase()) && field != null) {
+                        fieldApiNames.add(sourceFieldName.toLowerCase());
+                    }
 
-                        BDI_FieldMapping fieldMapping =
-                                fieldMappingByDevName.get(fieldMappingName);
+                    // Add the Name fields on the related object(s) for all Lookup fields,
+                    // used as display value(s) for Lookup fields loaded in the Form UI
+                    if ((fieldMapping != null &&
```

Comment:
I think we can extract the lines of code that populate `relatedObjectNameField` into two methods, `isRelatedObjectNameField` which returns a `boolean` and `getRelatedObjectNameField` that returns a `String`